src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceReader.scala

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2018 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.spark.sql.v2
+
+import java.util
+
+import com.mongodb.MongoClient
+import com.mongodb.client.{AggregateIterable, MongoCursor}
+import com.mongodb.spark.config.ReadConfig
+import com.mongodb.spark.rdd.partitioner.MongoPartition
+import com.mongodb.spark.sql.MapFunctions.documentToRow
+import com.mongodb.spark.sql.{MongoInferSchema, MongoRelationHelper}
+import com.mongodb.spark.sql.MongoRelationHelper.createPipeline
+import com.mongodb.spark.{LoggingTrait, MongoConnector, MongoSpark}
+import org.apache.spark.SparkContext
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.sources.{Filter, IsNotNull}
+import org.apache.spark.sql.sources.v2.reader.{DataReader, DataReaderFactory, DataSourceReader, SupportsPushDownFilters, SupportsPushDownRequiredColumns}
+import org.apache.spark.sql.types.StructType
+import org.bson.conversions.Bson
+import org.bson.{BsonArray, BsonDocument, BsonType}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+case class MongoDataSourceReader(private val schemaOption: Option[StructType], private val readConfig: ReadConfig) extends DataSourceReader
```

Comment:
Here we support pushed down filters and required columns through by implementing the interfaces.